Tests/Particles/SOAParticle/main.cpp

```diff
@@ -0,0 +1,178 @@
+    int const NArrayInt = pc.NArrayInt;
+
+    using ParticleType = typename T_PC::ParticleType;
+    using ParticleTileDataType = typename T_PC::ParticleTileType::ParticleTileDataType;
+    using RealVector = amrex::PODVector<ParticleReal, Allocator<ParticleReal> >;
+    using IntVector = amrex::PODVector<int, Allocator<int> >;
+    using SPType = typename T_PC::SuperParticleType;
+
+    const int add_num_particles = 5;
+
+    auto& ptile1 = pc.DefineAndReturnParticleTile(0, 0, 0);
+
+    for (int i = 0; i < add_num_particles; ++i)
+    {
+        ptile1.pos(i, 0) = 12.0;
+        ptile1.pos(i, 1) = 12.0;
+        ptile1.pos(i, 2) = 12.0;
+
+        // TODO
+        ptile1.id(i) = 1;
+        ptile1.cpu(i) = 1;
+    }
+    
+    int lev=0;
+    // int numparticles=0;
+    using MyParIter = ParIter_impl<ParticleType, NArrayReal, NArrayInt>;
+    for (MyParIter pti(pc, lev); pti.isValid(); ++pti) {
+        const int np = pti.numParticles();
+        //const auto t_lev = pti.GetLevel();
+        //const auto index = pti.GetPairIndex();
+        // ...
+
+        // preparing access to particle data: AoS
+        //using PType = ImpactXParticleContainer::ParticleType;
+        auto& aos = pti.GetArrayOfStructs();
+        ParticleType* AMREX_RESTRICT aos_ptr = aos().dataPtr();
+
+        // preparing access to particle data: SoA of Reals
+        auto& soa_real = pti.GetStructOfArrays().GetRealData();
+        amrex::ParticleReal* const AMREX_RESTRICT part_x = soa_real[0].dataPtr();
+        amrex::ParticleReal* const AMREX_RESTRICT part_y = soa_real[1].dataPtr();
+        amrex::ParticleReal* const AMREX_RESTRICT part_z = soa_real[2].dataPtr();
+        amrex::ParticleReal* const AMREX_RESTRICT part_aaa = soa_real[3].dataPtr();
+        auto& soa_int = pti.GetStructOfArrays().GetIntData();
+
+        // Iterating over old Particles
+        ParallelFor( np, [=] AMREX_GPU_DEVICE (long ip)
+        {
+            ParticleType& AMREX_RESTRICT p = aos_ptr[ip];
+            p.pos(0) += 1;
+            p.pos(1) += 1;
+            p.pos(2) += 1;
+            
+            amrex::ParticleReal & AMREX_RESTRICT x = part_x[ip];
+            amrex::ParticleReal & AMREX_RESTRICT y = part_y[ip];
+            amrex::ParticleReal & AMREX_RESTRICT z = part_z[ip];
+            amrex::ParticleReal & AMREX_RESTRICT a = part_aaa[ip];
+
+            x += 1.0;
+            y += 1.0;
+            z += 1.0;
+            a += 1.0;
+        });
+
+        // Iterating over SoA Particles
+        ParticleTileDataType ptd = pti.GetParticleTile().getParticleTileData(); 
+
+        ParallelFor( np, [=] AMREX_GPU_DEVICE (long ip)
+        {
+            ParticleType p(ptd, ip);
+            p.pos(0) += 1;
+            p.pos(1) += 1;
+            p.pos(2) += 1;
+        });
+
+
+    }
+
+    // create a host-side particle buffer
+    //ParticleContainer<1,1> pc_og;
+    auto tmp = pc.template make_alike<amrex::PinnedArenaAllocator>();
+    tmp.copyParticles(pc, true);
```

Comment:
This is the maybe last line currently failing @WeiqunZhang

Reply:
The issue is `amrex::ParticleTile<amrex::SoAParticle<3,4>, 3, 4, std::allocator>::ConstParticleTileDataType` returned by the call to `getConstParticelTileData` is `amrex::ConstParticleTileData<amrex::SoAParticleBase,3,4>`.  Note the extra `Base` in the name.  However, the lambda function defined in `addParticles` takes `amrex::::ConstParticleTileData<ParticleType, 3,4>`, where `ParticleType` is `amrex::SoAParticle<3,4>`, not `amrex::SoAParticleBase`.

Reply:
```
diff --git a/Src/Particle/AMReX_ParticleContainerI.H b/Src/Particle/AMReX_ParticleContainerI.H
index bf4b5d878c..51ef47a8b7 100644
--- a/Src/Particle/AMReX_ParticleContainerI.H
+++ b/Src/Particle/AMReX_ParticleContainerI.H
@@ -1000,8 +1000,7 @@ void
 ParticleContainer_impl<ParticleType, NArrayReal, NArrayInt, Allocator>::
 copyParticles (const PCType& other, bool local)
 {
-    using PData = ConstParticleTileData<ParticleType, NArrayReal, NArrayInt>;
-    copyParticles(other, [=] AMREX_GPU_HOST_DEVICE (const PData& /*data*/, int /*i*/) { return 1; }, local);
+    copyParticles(other, [=] AMREX_GPU_HOST_DEVICE (const auto& /*data*/, int /*i*/) { return 1; }, local);
 }

 template <typename ParticleType, int NArrayReal, int NArrayInt,
@@ -1011,8 +1010,7 @@ void
 ParticleContainer_impl<ParticleType, NArrayReal, NArrayInt, Allocator>::
 addParticles (const PCType& other, bool local)
 {
-    using PData = ConstParticleTileData<ParticleType, NArrayReal, NArrayInt>;
-    addParticles(other, [=] AMREX_GPU_HOST_DEVICE (const PData& /*data*/, int /*i*/) { return 1; }, local);
+    addParticles(other, [=] AMREX_GPU_HOST_DEVICE (const auto& /*data*/, int /*i*/) { return 1; }, local);
 }

 template <typename ParticleType, int NArrayReal, int NArrayInt,
```
This fixes it.  But I am not sure if it will have other consequences.

Reply:
I think this is a better fix than the generic lambda.
```
diff --git a/Src/Particle/AMReX_ParticleContainerI.H b/Src/Particle/AMReX_ParticleContainerI.H
index bf4b5d878c..098cdf47d7 100644
--- a/Src/Particle/AMReX_ParticleContainerI.H
+++ b/Src/Particle/AMReX_ParticleContainerI.H
@@ -1000,7 +1000,7 @@ void
 ParticleContainer_impl<ParticleType, NArrayReal, NArrayInt, Allocator>::
 copyParticles (const PCType& other, bool local)
 {
-    using PData = ConstParticleTileData<ParticleType, NArrayReal, NArrayInt>;
+    using PData = ConstParticleTileData<typename ParticleType::StorageParticleType, NArrayReal, NArrayInt>;
     copyParticles(other, [=] AMREX_GPU_HOST_DEVICE (const PData& /*data*/, int /*i*/) { return 1; }, local);
 }

@@ -1011,7 +1011,7 @@ void
 ParticleContainer_impl<ParticleType, NArrayReal, NArrayInt, Allocator>::
 addParticles (const PCType& other, bool local)
 {
-    using PData = ConstParticleTileData<ParticleType, NArrayReal, NArrayInt>;
+    using PData = ConstParticleTileData<typename ParticleType::StorageParticleType, NArrayReal, NArrayInt>;
     addParticles(other, [=] AMREX_GPU_HOST_DEVICE (const PData& /*data*/, int /*i*/) { return 1; }, local);
 }

```

Reply:
Thanks Weiqun on your comment, it seems good ! I will take a look at it and see if it does not have any consequences elsewhere